### PR TITLE
[SPARK-37613][SQL] Support ANSI Aggregate Function: regr_count

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
This PR used to support ANSI aggregate Function: `regr_count`

The mainstream database supports `regr_count` show below:
**Teradata**
https://docs.teradata.com/r/kmuOwjp1zEYg98JsB8fu_A/1skTm5SD0HD3Z~vii3hXAg
**Snowflake**
https://docs.snowflake.com/en/sql-reference/functions/regr_count.html#
**Oracle**
https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/REGR_-Linear-Regression-Functions.html#GUID-A675B68F-2A88-4843-BE2C-FCDE9C65F9A9
**H2**
http://www.h2database.com/html/functions-aggregate.html#regr_count
**Postgresql**
https://www.postgresql.org/docs/8.4/functions-aggregate.html
**Sybase**
https://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.help.sqlanywhere.12.0.0/dbreference/regr-count-function.html
**Exasol**
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/regr_function.htm

### Why are the changes needed?
`regr_count` is very useful.

### Does this PR introduce _any_ user-facing change?
'No'. New feature.

### How was this patch tested?
New tests.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @MaxGekk @gengliangwang cc @cloud-fan

**Comment by beliefer** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/AggregateExpressionSuite.scala [hunk 1]
```diff
@@ -31,4 +32,23 @@ class AggregateExpressionSuite extends SparkFunSuite {
     assert(expected == actual, s"Expected: $expected. Actual: $actual")
   }
 
+  test("test regr_count input types") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/AggregateExpressionSuite.scala:35]: This is too trivial and the type check framework should have already been well tested. I think we can remove it.

File: sql/core/src/test/resources/sql-tests/inputs/group-by.sql [hunk 2]
```diff
@@ -204,3 +207,13 @@ FROM VALUES
...
  (41), (42), (43), (44), (45), (46), (47), (48), (49), (50) AS tab(col);
+
+-- SPARK-37613: Support ANSI Aggregate Function: regr_count
+SELECT regr_count(b, a) FROM VALUES (1, 2L), (2, 2L), (2, 3L), (2, 4L) AS tab(a, b);
+SELECT regr_count(b, a) FROM VALUES (1, 2L), (2, null), (2, 3L), (2, 4L) AS tab(a, b);
+SELECT regr_count(b, a) FROM VALUES (1, 2L), (2, null), (null, 3L), (2, 4L) AS tab(a, b);
+SELECT a, regr_count(b, a) FROM VALUES (1, 2D), (2, 2L), (2, 3D), (2, 4D) AS tab(a, b) group by a;
+SELECT a, regr_count(b, a) FROM VALUES (1, 2D), (2, null), (2, 3D), (2, 4D) AS tab(a, b) group by a;
+SELECT a, regr_count(b, a) FROM VALUES (1, 2D), (2, null), (null, 3D), (2, 4D) AS tab(a, b) group by a;
+SELECT regr_count(y, x) FROM testRegression;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/group-by.sql:218]: what's special with `testRegression`?

**Review by cloud-fan — APPROVED** [review]
LGTM except for some minor comments.

File: sql/core/src/test/resources/sql-tests/inputs/group-by.sql (hunk 2 above)
Comment (beliefer) [reply sql/core/src/test/resources/sql-tests/inputs/group-by.sql:218]: The other Regression function will reuse `testRegression`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/AggregateExpressionSuite.scala (hunk 1 above)
Comment (beliefer) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/AggregateExpressionSuite.scala:35]: OK

File: sql/core/src/test/resources/sql-tests/inputs/group-by.sql (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/group-by.sql:218]: Then can we use it consistently and change `FROM VALUES...` to `FROM testRegression` in the tests above?

**Comment by beliefer** [comment]
retest this please

**Comment by cloud-fan** [comment]
please also address https://github.com/apache/spark/pull/34880#discussion_r770233966

File: sql/core/src/test/resources/sql-tests/inputs/group-by.sql (hunk 2 above)
Comment (beliefer) [reply sql/core/src/test/resources/sql-tests/inputs/group-by.sql:212]: OK

**Review by cloud-fan — APPROVED** [review]

**Comment by beliefer** [comment]
retest this please

**Comment by cloud-fan** [comment]
please also rerun the github actions (I don't have permissions to do so as the github action runs on your account)

**Comment by beliefer** [comment]
> please also rerun the github actions (I don't have permissions to do so as the github action runs on your account)

OK

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan Thanks a lot!

**Pull request closed without merging**